### ISSUE TITLE
CDLP test case for directed graphs

### ISSUE THREAD

**Issue by szarnyasg** [ldbc/ldbc_graphalytics#167]
@thegeman I am curious why this change was introduced: https://github.com/ldbc/ldbc_graphalytics/commit/48fdb8512df858187fedcbb90b1be53bd1a459e5.

If I created the test graph based on the input file ot the same commit:

https://github.com/ldbc/ldbc_graphalytics/blob/48fdb8512df858187fedcbb90b1be53bd1a459e5/graphalytics-validation/src/main/resources/validation-graphs/cdlp/dir-input#L1-L8

![image](https://user-images.githubusercontent.com/1402801/38769125-1db3a39a-3ffe-11e8-9d0a-c7ba752956ac.png)

As vertices 4-5 are connected both ways, why does v4 have a different partition label than v5?

I might miss something about the specification of the algorithm - feel free to correct me on the matter.

**Comment by thegeman** [comment]
Part of the input of a CDLP job is the number of iterations to perform. For the validation test case that number is 5:

https://github.com/ldbc/ldbc_graphalytics/blob/48fdb8512df858187fedcbb90b1be53bd1a459e5/graphalytics-validation/src/main/java/nl/tudelft/graphalytics/validation/algorithms/cdlp/CommunityDetectionLPValidationTest.java#L77

It turns out that this particular test case converges after iteration 6, so the label of v4 is still "incorrect" after iteration 5. I don't think the test case was deliberately designed this way, but it has helped find bugs in other platform implementations in the past.

**Comment by szarnyasg** [comment]
I see. This raises the interesting question whether implementations should follow the exact same algorithm that is presented in the specification.

It turns out that the Neo4j Graph Algorithms library uses a different notion of `iterations`. For 1 iteration, it does not yet converge:

```
CALL algo.labelPropagation(null, null, 'OUTGOING',
  {write: true, partitionProperty: 'LABEL', iterations: 1})
YIELD nodes, iterations, loadMillis, computeMillis, writeMillis, write, partitionProperty
```

Resulting graph:

```
╒════╤═══════╕
│"id"│"label"│
╞════╪═══════╡
│1   │2      │
├────┼───────┤
│2   │2      │
├────┼───────┤
│3   │2      │
├────┼───────┤
│4   │5      │
├────┼───────┤
│5   │5      │
├────┼───────┤
│6   │4      │
├────┼───────┤
│7   │5      │
├────┼───────┤
│8   │4      │
└────┴───────┘
```

Increasing the number of `iterations` to just 2 returns a correct result w.r.t. the communities detected (although with different labels).

```
CALL algo.labelPropagation(null, null, 'OUTGOING',
  {write: true, partitionProperty: 'LABEL', iterations: 2})
YIELD nodes, iterations, loadMillis, computeMillis, writeMillis, write, partitionProperty
```

Resulting graph:

```
╒════╤═══════╕
│"id"│"label"│
╞════╪═══════╡
│1   │2      │
├────┼───────┤
│2   │2      │
├────┼───────┤
│3   │2      │
├────┼───────┤
│4   │5      │
├────┼───────┤
│5   │5      │
├────┼───────┤
│6   │5      │
├────┼───────┤
│7   │5      │
├────┼───────┤
│8   │5      │
└────┴───────┘
```

**Comment by thegeman** [comment]
> This raises the interesting question of whether implementations should follow the exact same algorithm that is presented in the specification.

This has indeed been the topic of many discussions in the Graphalytics team. We would ideally leave all implementation details open so platforms can use whatever algorithm works best for them. Thus, Graphalytics requires that the output of each algorithm matches the reference implementation, but does not prescribe an implementation.

Exception to this rule is CDLP. To make validation of results possible, we use a deterministic definition of CDLP. Because this algorithm does not converge for all inputs, we opted to require a fixed number of iterations to be run. As a result, the reference (and required) output is intrinsically linked to our specific algorithm. As future work we may consider alternative definitions of what constitutes a "correct" implementation of CDLP (or even other CD algorithms), but this will not be included in Graphalytics v1.0.

**Comment by szarnyasg** [comment]
Thanks for the clarification - I am looking forward to discuss these issue in later Graphalytics versions.
